### PR TITLE
[Scatter2] Rewrite scatter plot tool in React

### PR DESCRIPTION
**Pull request by intrepiditee**
status: waiting for date and collection api

---
This change is [Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)

### CODE REVIEW

**Comment by intrepiditee** [comment]
While we wait for the new APIs, could you start reviewing? I don't anticipate large changes. Thanks!

I made some changes to `timeline_search.tsx` and a webdriver test kept failing for me and I could not fix it. Could you take a look?

@beets @shifucun @tjann

File: static/js/tools/scatter2/chart.tsx [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Chart component for retrieving, transforming, and plotting data.
```
Comment (beets) [inline static/js/tools/scatter2/chart.tsx:18]: This component might be trying to do too much - I think it'll be easier to manage if the data loading happened elsewhere (in another class perhaps), then this component can focus on the presentation layer.

**Review by beets — COMMENTED** [review]
Sorry for the delay with this! Learning a lot about modern react from this PR.

Overall it's well laid out and thanks for the tests and comments! I mostly looked at chart & context - didn't look closely at the other modules or the drawing code in chart.

**Review by shifucun — CHANGES_REQUESTED** [review]
> Reviewed 1 of 19 files at r1.
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)* status: 1 of 19 files reviewed, 24 unresolved discussions (waiting on @intrepiditee and @shifucun)

___
*[server/webdriver_tests/screenshot/screenshot_test.py, line 112 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MMvlUz74xbP9GW-NCCo:-MMvlUz74xbP9GW-NCCp:b-xmmr3y) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/server/webdriver_tests/screenshot/screenshot_test.py#L112)):*
> ```Python
>         'height': 2300
>     },
>     {
> ```

Can you make the url non-encoded so it's easier to update later.
___
*[static/css/timeline.scss, line 17 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MMvmICy26S5LMdShy0D:-MMvmICy26S5LMdShy0E:b-l9n9kf) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/css/timeline.scss#L17)):*
> ```SCSS
>  */
> 
>  @use "search_place";
> ```

extra space
___
*[static/js/tools/scatter2/app.tsx, line 91 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0hAkwB1bWfPJta-TQ:-MN0hAkwB1bWfPJta-TR:bfeu6vk) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/app.tsx#L91)):*
> ```XML
>  * @param context
>  */
> function applyHash(context: ContextType) {
> ```

I found it a bit hard to understand the usage of store here: the context(store) is provided but setAxis and setPlace also needs to be imported here, that seems redundent. Should the store wraps all these instead it? For example:

context.setAxis()
context.setPlace()
___
*[static/js/tools/scatter2/chart.tsx, line 541 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0hnor-GUjZFcsY8e-:-MN0hnor-GUjZFcsY8e0:b-s9uj57) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/chart.tsx#L541)):*
> ```XML
>  * @param place
>  */
> function loadData(axis: ContextFieldType<Axis>, place: Place): void {
> ```

Could be have another module that has a custom hook for data loading? Could be use anything from this post: https://www.robinwieruch.de/react-hooks-fetch-data?
___
*[static/js/tools/scatter2/context.tsx, line 26 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0P3C0CZbsBkEjVczz:-MN0P3C0CZbsBkEjVd--:b-e4cgki) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L26)):*
> ```XML
> type Setter<T> = React.Dispatch<React.SetStateAction<T>>;
> 
> interface Axis {
> ```

Is it better to separate out population and perCapita out of Axis class struct. As I see they should be part of data massage logic instead of UI logic.  The Axis should take ready to use data to plot as a generic interface..
___
*[static/js/tools/scatter2/context.tsx, line 55 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0RsEr4jeREm60QRCd:-MN0RsEr4jeREm60QRCe:b-ru3jir) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L55)):*
> ```XML
> }
> 
> interface Place {
> ```

Naming suggestion: how about PlaceInfo for this interface. "Place" is too generic regarding the function of this interface.
___
*[static/js/tools/scatter2/context.tsx, line 78 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0UIST5QTKwRZYm-dy:-MN0UIST5QTKwRZYm-dz:bk10e99) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L78)):*
> ```XML
> });
> 
> interface ContextFieldType<V> {
> ```

I feel "StateType" or "useStateType" or something similar is easier to understand here. Since it's to wrap a state and setState type and has nothing to do with the context (it just happend that the context only contains the state?)
___
*[static/js/tools/scatter2/context.tsx, line 98 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0TEUI28s1ICLZl20f:-MN0TEUJ0xJUd9Q5SpF8:bqe6jve) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L98)):*
> ```XML
>  * Constructs an initial context.
>  */
> function useStore(): ContextType {
> ```

might be clear to call it useStateStore or useContextStore ?
___
*[static/js/tools/scatter2/context.tsx, line 102 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0Sg5jDLV_pTWn3I0h:-MN0Sg5jDLV_pTWn3I0i:bmagmmf) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L102)):*
> ```XML
>   const [y, setY] = useState(EmptyAxis);
>   const [place, setPlace] = useState({
>     enclosingPlace: {
> ```

use EmptyPlace here?
___
*[static/js/tools/scatter2/context.tsx, line 130 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0W1252zKxkPItB7Wq:-MN0W1252zKxkPItB7Wr:b-p4mln8) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L130)):*
> ```XML
>  * @param field
>  * @param place
>  */
> ```

curious is the "set***" pattern here a popular way of doing thing when using react hook? I find it a bit hard to understand and cumbersome than using a "Class store". If class is not possible, could we make "setX, setY, setPlace" as module level variable and export them? That way can save all the argument of "field: ContextFieldType"?
___
*[static/js/tools/scatter2/context.tsx, line 229 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0Wyyb0Eptt5MwEVso:-MN0Wyyc386EBvfdH1CM:bj05ivz) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/context.tsx#L229)):*
> ```XML
>   place.set({ ...place.value, upperBound: bound });
> }
> ```

another question as I am new to the react hook usage. Looks like all the method here is trying to manipulate the nested store fields. Would some reducer help here either for code or logic cleanness?
___
*[static/js/tools/scatter2/place_options.tsx, line 31 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0eg-I07-KFiBEOSvZ:-MN0eg-I07-KFiBEOSv_:bxhnov2) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/place_options.tsx#L31)):*
> ```
>   setEnclosedPlaces,
>   setEnclosedPlaceType,
>   setEnclosingPlace,
> ```
Similar question here, I feel the import list here can easily grow to very long if the context fields increases. Is there any other way/pattern on this?

Is it possible to import a "store" instance from context.ts and use it everywhere? Sorry my knowledge on Hook is very limited and this might be a naive question.
___
*[static/js/tools/scatter2/place_options.tsx, line 142 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0fOjV5qsQ3rSe7Deo:-MN0fOjV5qsQ3rSe7Dep:boi4ffh) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/place_options.tsx#L142)):*
> ```
>     if (!_.isEmpty(dcids)) {
>       getPlaceNames(dcids).then((dcidToName) => {
>         setEnclosedPlaces(
> ```
Might be easier to view if you do sequential doAsync(() => {return x}).then((data) => {return y}).then((data) => {return z}). nesting "then()" is a bit hard to read.
___
*[static/js/tools/scatter2/statvar.tsx, line 100 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0g32NCrhxlZQ3bK5d:-MN0g32O5DnrvMOPSDB7:b-58bewj) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/statvar.tsx#L100)):*
> ```XML
>     setLoading(true);
>   }, [context.place.value.enclosedPlaces]);
> ```

Add a comment for this useEffect()?
___
*[static/js/tools/scatter2/util.ts, line 29 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0gR-uAfohTvX2QyHv:-MN0gR-uAfohTvX2QyHw:bahyhs6) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/util.ts#L29)):*
> ```XML
>  * @param type
>  */
> async function getPlacesIn(dcid: string, type: string): Promise<Array<string>> {
> ```

As discussed, can you move these api call to the flask server?

**Review by shifucun — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)* status: 1 of 19 files reviewed, 25 unresolved discussions (waiting on @intrepiditee and @shifucun)

*[a discussion](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0iOfk3Z007DpsG8me:-MN0iOfk3Z007DpsG8mf:b-a49h3k) (no related file):*
Very clean and structured code. My comments are mostly for the use of React hook and make the wrapper/abstraction easy to understand.

---

File: static/css/tools/scatter2.scss [hunk 2]
```diff
@@ -14,8 +14,76 @@
...
+
+@include scatter-search;
+
+#main-pane #plot-container {
+  direction: initial;
+  padding: 2rem;
+  overflow-y: scroll;
+}
 
 // TODO(intrepiditee)
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/css/tools/scatter2.scss:42]: forgot to remove this

File: static/js/tools/scatter2/app.test.tsx [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+import { App } from "./app";
+import { Context, setEnclosingPlace, useStore } from "./context";
+
+import hierarchy from "../../../data/hierarchy_top.json";
+import { waitFor } from "@testing-library/react";
+import { getApiKey, getApiRoot } from "../../shared/util";
+
+Enzyme.configure({ adapter: new Adapter() });
+
+test("Delaware Number Of Establishments vs employed", async (done) => {
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/app.test.tsx:32]: fixed

File: static/js/tools/scatter2/app.test.tsx [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    expect(text).toContain("Number Of Establishments vs Employed");
+    // Stats
+    expect(text).toContain("X Mean: 152696");
+    expect(text).toContain("Y Mean: 8359.667");
+    expect(text).toContain("X Standard Deviation: 108149.894");
+    expect(text).toContain("Y Standard Deviation: 6753.678");
+  });
+
+  // Points
+  const expectThreeCircles = () => {
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/app.test.tsx:67]: fixed

File: static/js/tools/scatter2/context.tsx [hunk 5]
```diff
@@ -0,0 +1,255 @@
+/**
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/context.tsx:1]: fixed

File: static/js/tools/scatter2/chart.tsx [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+ * Formats a number, or returns "N/A" if not an number.
+ * If the number is a float, keeps three decimal places.
+ * @param num
+ */
+function getStringOrNA(num: number): string {
+  return _.isNil(num)
+    ? "N/A"
+    : Number.isInteger(num)
+    ? num.toString()
+    : num.toFixed(3);
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/chart.tsx:256]: added

File: static/js/tools/scatter2/chart.tsx [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+  xPop: number;
+  yPop: number;
+  place: NamedPlace;
+}
+
+// TODO: Show provenance.
+function Chart(): JSX.Element {
+  const context = useContext(Context);
+  const [points, setPoints] = useState([] as Array<Point>);
+  const [loading, setLoading] = useState(false);
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/chart.tsx:53]: fixed

File: static/js/tools/scatter2/chart.tsx [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+    if (downloadButton) {
+      downloadButton.style.visibility = "visible";
+      downloadButton.onclick = () => downloadData(x.value, y.value, place);
+    }
+  }, [context]);
+
+  // Replot when data changes.
+  useEffect(() => {
+    if (
+      !isStatVarNameLoaded(context.x.value) ||
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/chart.tsx:92]: In `StatVarChooser` in `statvar.tsx`. The names are passed by the statvar menu.

File: static/js/tools/scatter2/chart.tsx [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+    const x = context.x;
+    const y = context.y;
+    if (!isStatVarNameLoaded(x.value) || !isStatVarNameLoaded(y.value)) {
+      return;
+    }
+    setLoading(true);
+    loadPopulationsAndDataIfNeeded(x, place);
+    loadPopulationsAndDataIfNeeded(y, place);
+    if (!arePopulationsAndDataLoaded(x.value, y.value)) {
+      return;
```
[the remark this answers is not in the document]
Comment (intrepiditee) [reply static/js/tools/scatter2/chart.tsx:75]: This is not an error state. The `loadPopulationsAndDataIfNeeded` calls are async, so when `arePopulationsAndDataLoaded` is called, the data may not have been loaded. When they are loaded, this hook will get called again and that's when `arePopulationsAndDataLoaded` will return `true` and the plotting happens.

File: static/js/tools/scatter2/chart.tsx (hunk 1 above)
Comment (intrepiditee) [reply static/js/tools/scatter2/chart.tsx:18]: separated into `ChartLoader` and `Chart`

**Comment by intrepiditee** [comment]
> > _[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)_ status: 1 of 19 files reviewed, 25 unresolved discussions (waiting on @intrepiditee and @shifucun)
> 
> _[a discussion](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0iOfk3Z007DpsG8me:-MN0iOfk3Z007DpsG8mf:b-a49h3k) (no related file):_
> Very clean and structured code. My comments are mostly for the use of React hook and make the wrapper/abstraction easy to understand.

Thanks for reviewing.

For the context, the `setXXX` functions are totally unnecessary. I just wanted to place all the logic that manipulates the store in one file. If we want, we can remove and inline all those functions. Another way that I think is nicer is just like what you said - attaching the functions to the store so we can then do `context.setXXX` instead. Let me know which way you prefer. Thanks.

**Review by shifucun — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)* status: 0 of 20 files reviewed, 25 unresolved discussions (waiting on @beets, @intrepiditee, and @shifucun)

*[a discussion](https://reviewable.io/reviews/datacommonsorg/website/668#-MN0iOfk3Z007DpsG8me:-MNG4Y2M1pWA9rdIugYY:blnwgqx) (no related file):*
<details><summary><i>Previously, intrepiditee wrote…</i></summary><blockquote>

Thanks for reviewing.

For the context, the `setXXX` functions are totally unnecessary. I just wanted to place all the logic that manipulates the store in one file. If we want, we can remove and inline all those functions. Another way that I think is nicer is just like what you said - attaching the functions to the store so we can then do `context.setXXX` instead. Let me know which way you prefer. Thanks. 
</blockquote></details>

attaching the functions to the store would be good!

---

**Comment by intrepiditee** [comment]
@beets @shifucun 

Thanks for reviewing. Most of the comments have been addressed. An exception is moving `populations` and `perCapita` out of `Axis`, mainly because I am not yet sure if that's better and whether we should also move `data` and `log` out. We can discuss more on Monday, but I think this PR is ready for another round of reviews. Thanks.

**Review by shifucun — APPROVED** [review]
Thanks for the update!

**Comment by intrepiditee** [comment]
> Thanks for the update!

Hi Bo, I am thinking about pulling `populations` and `data` out of `Axis` but keeping `log` and `perCapita`. What do you think and is it a good idea? Thanks

**Comment by shifucun** [comment]
> > Thanks for the update!
> 
> Hi Bo, I am thinking about pulling `populations` and `data` out of `Axis` but keeping `log` and `perCapita`. What do you think and is it a good idea? Thanks

perCapita is more of a data processing step.. Imagine the case 1) a special denominator other than total Population 2) other operations than "perCapita" are cases that can blow off this data manipulation.  Have this step separate out will be simple to maintain in general. The the "Axis" is really just about plot.

**Comment by intrepiditee** [comment]
> > > Thanks for the update!
> > 
> > 
> > Hi Bo, I am thinking about pulling `populations` and `data` out of `Axis` but keeping `log` and `perCapita`. What do you think and is it a good idea? Thanks
> 
> perCapita is more of a data processing step.. Imagine the case 1) a special denominator other than total Population 2) other operations than "perCapita" are cases that can blow off this data manipulation. Have this step separate out will be simple to maintain in general. The the "Axis" is really just about plot.

Could you explain a bit more what's the issue with other denominators? `perCapita` is a boolean just like `log`

**Comment by shifucun** [comment]
> > > > Thanks for the update!
> > > 
> > > 
> > > Hi Bo, I am thinking about pulling `populations` and `data` out of `Axis` but keeping `log` and `perCapita`. What do you think and is it a good idea? Thanks
> > 
> > 
> > perCapita is more of a data processing step.. Imagine the case 1) a special denominator other than total Population 2) other operations than "perCapita" are cases that can blow off this data manipulation. Have this step separate out will be simple to maintain in general. The the "Axis" is really just about plot.
> 
> Could you explain a bit more what's the issue with other denominators? `perCapita` is a boolean just like `log`

I see, so for the Context struct, perCapita is fine. I was thinking it for plot but it's actually a global state. That said, maybe the perCapita field in chart.tsx can be taken out. The plot params is derived from it.

**Comment by intrepiditee** [comment]
That makes sense. Will do that. Thanks.

**Review by beets — CHANGES_REQUESTED** [review]
> Reviewed 3 of 19 files at r1, 1 of 1 files at r3, 2 of 11 files at r4, 1 of 14 files at r5, 6 of 7 files at r6, 2 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)* status: 15 of 26 files reviewed, 18 unresolved discussions (waiting on @beets and @intrepiditee)

___
*[server/routes/api/place.py, line 546 at r6](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_Wlgq9RVl1ozPLtxl:-MN_Wlgq9RVl1ozPLtxm:b-oh85ul) ([raw file](https://github.com/datacommonsorg/website/blob/9f0237cfffc923d654bfd64fe170f2c2b333e4e3/server/routes/api/place.py#L546)):*
> ```Python
> 
> @bp.route('/places-in')
> @cache.cached(timeout=3600 * 24, query_string=True)  # Cache for one day.
> ```

This won't work for get arguments since it won't be in the cache key -- we've worked around this in the past by adding the args to the path, or calling a helper that is cached.
___
*[server/routes/api/stats.py, line 138 at r6](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_XVzx39a0Mlv8R8Lr:-MN_XVzx39a0Mlv8R8Ls:bocu7gp) ([raw file](https://github.com/datacommonsorg/website/blob/9f0237cfffc923d654bfd64fe170f2c2b333e4e3/server/routes/api/stats.py#L138)):*
> ```Python
> 
> @bp.route('/api/stats/value')
> @cache.cached(timeout=3600 * 24, query_string=True)  # Cache for one day.
> ```

There are other examples of how we work around the cache limitation with get params in this file - see get_stats_wrapper.
___
*[static/js/tools/scatter2/chart.tsx, line 92 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MMvh0GFs8vLPjhoHrwW-r2-92:-MN__gnr-fBpCSyyCoWq:b-xe95w9) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/chart.tsx#L92)):*
<details><summary><i>Previously, intrepiditee wrote…</i></summary><blockquote>

In `StatVarChooser` in `statvar.tsx`. The names are passed by the statvar menu.
</blockquote></details>

Gotcha - thanks

**Comment by intrepiditee** [comment]
___
*[server/routes/api/place.py, line 546 at r6](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_Wlgq9RVl1ozPLtxl:-MN_ha7CBZZmOngJ4v8h:b-pxwj1n) ([raw file](https://github.com/datacommonsorg/website/blob/9f0237cfffc923d654bfd64fe170f2c2b333e4e3/server/routes/api/place.py#L546)):*
<details><summary><i>Previously, beets wrote…</i></summary><blockquote>

This won't work for get arguments since it won't be in the cache key -- we've worked around this in the past by adding the args to the path, or calling a helper that is cached.
</blockquote></details>

I thought the `query_string` option was for that?

**Review by beets — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)* status: 15 of 26 files reviewed, 17 unresolved discussions (waiting on @intrepiditee)

___
*[server/routes/api/place.py, line 546 at r6](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_Wlgq9RVl1ozPLtxl:-MN_pQuZ5DKyOzlSfiWL:b-bqia0v) ([raw file](https://github.com/datacommonsorg/website/blob/9f0237cfffc923d654bfd64fe170f2c2b333e4e3/server/routes/api/place.py#L546)):*
<details><summary><i>Previously, intrepiditee wrote…</i></summary><blockquote>

I thought the `query_string` option was for that?
</blockquote></details>

You're right! And I missed that you added that option. Thanks, that's good to know.
___
*[static/js/tools/scatter2/chart.tsx, line 18 at r2](https://reviewable.io/reviews/datacommonsorg/website/668#-MMvh0GFs8vLPjhoHrwW-r2-18:-MN_pjd0151tsRIgk6lX:b-tc46s) ([raw file](https://github.com/datacommonsorg/website/blob/093561d3c1eb0b42f8a155053a4fd1439c967e72/static/js/tools/scatter2/chart.tsx#L18)):*
<details><summary><i>Previously, intrepiditee wrote…</i></summary><blockquote>

separated into `ChartLoader` and `Chart`
</blockquote></details>

Thanks, those changes look good

**Review by tjann — APPROVED** [review]
> Reviewed 2 of 19 files at r1, 6 of 14 files at r5, 1 of 7 files at r6, 1 of 3 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/website/668)* status: all files reviewed, 20 unresolved discussions (waiting on @intrepiditee)

___
*[static/js/tools/scatter2/chart.tsx, line 37 at r8](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_tJZl8TsHrHjVYTZT:-MN_tJZl8TsHrHjVYTZU:b-18xc90) ([raw file](https://github.com/datacommonsorg/website/blob/da75abfc699c4c2b2e1a84c2567c5dea6aa46c51/static/js/tools/scatter2/chart.tsx#L37)):*
> ```XML
> }
> 
> // TODO: Show provenance.
> ```

Is there a github issue for this? Might be good to have a reference
___
*[static/js/tools/scatter2/chart.tsx, line 42 at r8](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_tT3SEi5ThTWpCiT7:-MN_tT3SEi5ThTWpCiT8:b-a09w3v) ([raw file](https://github.com/datacommonsorg/website/blob/da75abfc699c4c2b2e1a84c2567c5dea6aa46c51/static/js/tools/scatter2/chart.tsx#L42)):*
> ```XML
>   const tooltip = useRef<HTMLDivElement>();
> 
>   // Replot when data changes.
> ```

Thanks for the comments!
___
*[static/js/tools/scatter2/chart.tsx, line 74 at r8](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_tvYX2257AjJ-IDld:-MN_tvYX2257AjJ-IDle:b-qew3lt) ([raw file](https://github.com/datacommonsorg/website/blob/da75abfc699c4c2b2e1a84c2567c5dea6aa46c51/static/js/tools/scatter2/chart.tsx#L74)):*
> ```XML
>  * Formats a number, or returns "N/A" if not an number.
>  * If the number is a float, keeps three decimal places.
>  * TODO: Three decimal places may not be enough. Need a utility to do this.
> ```

Nit: I'm not sure what we need a utility to do? More than 3 decimals?
___
*[static/js/tools/scatter2/statvar.tsx, line 204 at r8](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_usBW1RYh-p5bis3o:-MN_usBW1RYh-p5bis3p:bu7hbu9) ([raw file](https://github.com/datacommonsorg/website/blob/da75abfc699c4c2b2e1a84c2567c5dea6aa46c51/static/js/tools/scatter2/statvar.tsx#L204)):*
> ```XML
>  * Filters statvars shown in the statvar menu.
>  * A statvar is kept if it is available for at least one of the child places.
>  * Throws an alert if a currently selected statvar is fitlered out.
> ```

nit: filtered

**Comment by intrepiditee** [comment]
___
*[static/js/tools/scatter2/chart.tsx, line 74 at r8](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_tvYX2257AjJ-IDld:-MNdi_XF2ohunYUpGdJY:bu5bnmp) ([raw file](https://github.com/datacommonsorg/website/blob/da75abfc699c4c2b2e1a84c2567c5dea6aa46c51/static/js/tools/scatter2/chart.tsx#L74)):*
<details><summary><i>Previously, tjann (TJann) wrote…</i></summary><blockquote>

Nit: I'm not sure what we need a utility to do? More than 3 decimals?
</blockquote></details>

to determine the number of decimals places based on value

**Comment by intrepiditee** [comment]
___
*[static/js/tools/scatter2/statvar.tsx, line 204 at r8](https://reviewable.io/reviews/datacommonsorg/website/668#-MN_usBW1RYh-p5bis3o:-MNdihKZ7285Ek5Z5Ctl:b-7ql1td) ([raw file](https://github.com/datacommonsorg/website/blob/da75abfc699c4c2b2e1a84c2567c5dea6aa46c51/static/js/tools/scatter2/statvar.tsx#L204)):*
<details><summary><i>Previously, tjann (TJann) wrote…</i></summary><blockquote>

nit: filtered
</blockquote></details>

fixed

**Comment by intrepiditee** [comment]
@shifucun @beets @tjann I am happy with the pull request. Could you merge? I don't have write access.

**Comment by beets** [comment]
thanks @intrepiditee! merged, and we should see it on staging soon!

**Pull request merged**